vault/resource_secrets_sync_association.go

```diff
@@ -89,48 +106,99 @@ func secretsSyncAssociationWrite(ctx context.Context, d *schema.ResourceData, me
-	accessor, err := getMountAccessor(ctx, d, meta)
-	if err != nil {
-		return diag.Errorf("could not obtain accessor from given mount; err=%s", err)
+	// ex: gh/dest/gh-dest-1/mount/kv/secret/token
+	// unique. each destination should only have one association for a particular accessor and secret
+	id := fmt.Sprintf("%s/dest/%s/mount/%s/secret/%s", destType, name, mount, secretName)
+	d.SetId(id)
+
+	return secretsSyncAssociationRead(ctx, d, meta)
+}
+
+func secretsSyncAssociationRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
 	}
-	vaultRespKey := fmt.Sprintf("%s/%s", accessor, secretName)
 
-	saModel, err := getSyncAssociationModelFromResponse(resp)
+	id := d.Id()
+	fields, err := syncAssociationFieldsFromID(id)
 	if err != nil {
 		return diag.FromErr(err)
 	}
 
-	syncData, ok := saModel.AssociatedSecrets[vaultRespKey]
-	if !ok {
-		return diag.Errorf("no associated secrets found for given mount accessor and secret name %s", vaultRespKey)
+	typ := fields[0]
+	destName := fields[1]
+	mount := fields[2]
+	secretName := fields[3]
+
+	if err := d.Set(fieldSecretName, secretName); err != nil {
+		return diag.FromErr(err)
 	}
 
-	// set data that is received from Vault upon writes to avoid extra sync association reads
-	if err := d.Set(fieldSecretName, syncData.SecretName); err != nil {
+	if err := d.Set(consts.FieldName, destName); err != nil {
 		return diag.FromErr(err)
 	}
 
-	if err := d.Set(fieldSyncStatus, syncData.SyncStatus); err != nil {
+	if err := d.Set(consts.FieldType, typ); err != nil {
 		return diag.FromErr(err)
 	}
 
-	if err := d.Set(fieldUpdatedAt, syncData.UpdatedAt); err != nil {
+	if err := d.Set(consts.FieldMount, mount); err != nil {
 		return diag.FromErr(err)
 	}
 
-	d.SetId(path)
+	accessor, err := getMountAccessor(ctx, d, meta, mount)
+	if err != nil {
+		return diag.Errorf("could not obtain accessor from given mount; err=%s", err)
+	}
+	// List all associations for secret destination
+	resp, err := client.Logical().Read(fmt.Sprintf("%s/%s", syncutil.SecretsSyncDestinationPath(destName, typ), "associations"))
+	if err != nil {
+		return diag.Errorf("error reading associations for destination %s of type %s", destName, typ)
+	}
 
-	return nil
-}
+	model, err := getSyncAssociationModelFromResponse(resp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	status := map[string]string{}
+	subKeys := make([]string, 0)
+	var updatedAt string
+	prefix := fmt.Sprintf("%s/%s", accessor, secretName)
+	for k, v := range model.AssociatedSecrets {
+		if strings.HasPrefix(k, prefix) {
+			status[k] = v.SyncStatus
+			// only add sub-keys if they are non-zero
+			if v.Subkey != "" {
+				subKeys = append(subKeys, v.Subkey)
+			}
+			updatedAt = v.UpdatedAt
+		}
+	}
+
+	if len(status) == 0 {
+		return diag.Errorf("no associated secrets found for given mount accessor and secret name %s/%s", accessor, secretName)
+	}
```

Comment:
This is the main bit of changes for the Association resource. the `sync_status` is now a map that stores sync statuses for all subkeys. Here is what it looks like in TF state:
```
KV Data:
{
      dev  = "B!gS3cr3t",
      prod = "S3cureP4$$"
}

TF state:
"sync_status": {
              "kv_32f9435a/token/dev": "SYNCED",
              "kv_32f9435a/token/prod": "SYNCED"
}
```

Reply:
With latest commit, `updated_at` is also a map. Ex:
```
"updated_at": {
       "kv_201da4fd/token/dev": "2024-03-21T12:42:02.558533-07:00",
       "kv_201da4fd/token/prod": "2024-03-21T12:42:02.795474-07:00"
}
```

Reply:
Instead of having N different maps in the TF state, where N is number of different fields, what do you think about having a single type to hold everything?
```go 
type Associations map[string]struct{
    SyncStatus string
    UpdatedAt time.Time
    //.. future fields here
}
```

Reply:
@raymonstah, I thought about that, but due to the nature of the TF SDKv2, I don't think this would be possible for maps. For map schema types, the SDK only offers a `TypeMap`, which is a `map[string]string`. To be able to store all the metadata in this form in a single type, we will have to `json.Stringify` the struct and store it as a string, which I'm not sure would be as helpful to the end user (since they can't loop over any subkeys/metadata)

Alternatively we can store these as a `TypeList` instead of ` TypeMap`. This wouldn't allow us to map the metadata to the particular subkey, but we could store the particular subkey along with all the metadata in the list. Curious to hear what folks think would be a better UX for the user to be able to use the data in these computed fields

Reply:
Ah I see. Thanks for that context. No objections to `TypeList`. The solution to store the subkey along with the metadata makes sense to me.

Reply:
Example:
```
"metadata": [
              {
                "sub_key": "prod",
                "sync_status": "SYNCED",
                "updated_at": "2024-03-25T11:45:22.603164-07:00"
              },
              {
                "sub_key": "dev",
                "sync_status": "SYNCED",
                "updated_at": "2024-03-25T11:45:22.383758-07:00"
              }
],
```